src/coprocessor/dag/rpn_expr/impl_arithmetic.rs

```diff
@@ -4,74 +4,110 @@ use cop_codegen::RpnFunction;
 
 use super::types::RpnFnCallPayload;
 use crate::coprocessor::codec::data_type::*;
+use crate::coprocessor::codec::mysql::Res;
 use crate::coprocessor::codec::{self, Error};
 use crate::coprocessor::dag::expr::EvalContext;
+use crate::coprocessor::dag::rpn_expr::Uint;
 use crate::coprocessor::Result;
 use std::fmt::Debug;
 
 #[derive(Debug, RpnFunction)]
 #[rpn_function(args = 2)]
-pub struct RpnFnArithmetic<A: ArithmeticOp> {
-    _phantom: std::marker::PhantomData<A>,
+pub struct RpnFnArithmetic<Arg0, Arg1, Ret, Op>
+where
+    Arg0: Evaluable,
+    Arg1: Evaluable,
+    Ret: Evaluable,
+    Op: ArithmeticOp<Arg0, Arg1, Ret>,
+{
+    _phantom: std::marker::PhantomData<(Arg0, Arg1, Ret, Op)>,
 }
 
-impl<A: ArithmeticOp> RpnFnArithmetic<A> {
-    pub fn new() -> Self {
+impl<Arg0, Arg1, Ret, Op> RpnFnArithmetic<Arg0, Arg1, Ret, Op>
+where
+    Arg0: Evaluable,
+    Arg1: Evaluable,
+    Ret: Evaluable,
+    Op: ArithmeticOp<Arg0, Arg1, Ret>,
+{
+    fn new() -> Self {
         Self {
             _phantom: std::marker::PhantomData,
         }
     }
 
     #[inline]
     fn call(
-        _ctx: &mut EvalContext,
+        ctx: &mut EvalContext,
         _payload: RpnFnCallPayload<'_>,
-        arg0: &Option<A::T>,
-        arg1: &Option<A::T>,
-    ) -> Result<Option<A::T>> {
+        arg0: &Option<Arg0>,
+        arg1: &Option<Arg1>,
+    ) -> Result<Option<Ret>> {
         if let (Some(lhs), Some(rhs)) = (arg0, arg1) {
-            A::calc(lhs, rhs)
+            Op::calc(ctx, lhs, rhs)
         } else {
             // All arithmetical functions with a NULL argument return NULL
             Ok(None)
         }
     }
 }
 
-impl<A: ArithmeticOp> Clone for RpnFnArithmetic<A> {
+impl<Arg0, Arg1, Ret, Op> Clone for RpnFnArithmetic<Arg0, Arg1, Ret, Op>
+where
+    Arg0: Evaluable,
+    Arg1: Evaluable,
+    Ret: Evaluable,
+    Op: ArithmeticOp<Arg0, Arg1, Ret>,
+{
     fn clone(&self) -> Self {
         Self::new()
     }
 }
 
-impl<A: ArithmeticOp> Copy for RpnFnArithmetic<A> {}
-
-pub trait ArithmeticOp: Send + Sync + Debug + 'static {
-    type T: Evaluable;
+impl<Arg0, Arg1, Ret, Op> Copy for RpnFnArithmetic<Arg0, Arg1, Ret, Op>
+where
+    Arg0: Evaluable,
+    Arg1: Evaluable,
+    Ret: Evaluable,
+    Op: ArithmeticOp<Arg0, Arg1, Ret>,
+{
+}
 
-    fn calc(lhs: &Self::T, rhs: &Self::T) -> Result<Option<Self::T>>;
+pub trait ArithmeticOp<Arg0, Arg1 = Arg0, Ret = Arg0>: Send + Sync + Debug + 'static
+where
+    Arg0: Evaluable,
+    Arg1: Evaluable,
+    Ret: Evaluable,
+{
+    fn calc(ctx: &mut EvalContext, lhs: &Arg0, rhs: &Arg1) -> Result<Option<Ret>>;
+
+    fn func() -> RpnFnArithmetic<Arg0, Arg1, Ret, Self>
+    where
+        Self: Sized,
+    {
+        RpnFnArithmetic::new()
+    }
 }
 
-#[derive(Debug)]
-pub struct IntIntPlus;
+macro_rules! arith_op {
```

Comment:
Looks like this macro does not save much code 😂

Reply:
Right. I create this macro just because lots of `PhantomData` looks bad...